runtime/src/main/java/org/corfudb/runtime/LogReplicationListener.java

```diff
@@ -0,0 +1,182 @@
+ * received through replication and local updates.  Log Replicator does not write to this merged table.  This
+ * listener will observe the writes and apply them to the merged table based on the client implementation.
+ *
+ *
+ * This interface sees ordered updates from :
+ * 1. client-streams from client-Namespace, and,
+ * 2. LrStatusTable from corfuSystem-Namespace.
+ *
+ * The client implementing this interface will only observe the data updates from client streams
+ */
+@Slf4j
+public abstract class LogReplicationListener implements StreamListener {
+
+    // Indicates if a full sync on client tables was performed during subscription.  A full sync will not be
+    // performed if a snapshot sync is ongoing.
+    @Getter
+    private final AtomicBoolean clientFullSyncPending = new AtomicBoolean(false);
+
+    // This variable tracks if a snapshot sync is ongoing
+    @Getter
+    private final AtomicBoolean snapshotSyncInProgress = new AtomicBoolean(false);
+
+    // Timestamp at which the client performed a full sync.  Any updates below this timestamp must be ignored.
+    // At the time of subscription, a full sync cannot be performed if LR Snapshot Sync is in progress.  Full Sync is
+    // performed when this ongoing snapshot sync completes.  The listener, however, can get updates before this full
+    // sync.  So we need to maintain this timestamp and ignore any updates below it.
+    @Getter
+    private final AtomicLong clientFullSyncTimestamp = new AtomicLong(Address.NON_ADDRESS);
+
+    private final CorfuStore corfuStore;
+    private final String namespace;
+
+    /**
+     * Special LogReplication listener which a client creates to receive ordered updates for replicated data.
+     * @param corfuStore Corfu Store used on the client
+     * @param namespace Namespace of the client's tables
+     */
+    public LogReplicationListener(CorfuStore corfuStore, @Nonnull String namespace) {
+        this.corfuStore = corfuStore;
+        this.namespace = namespace;
+    }
+
+    /**
+     * This is an internal method of this abstract listener and not exposed to clients.
+     *
+     * @param results is a map of stream UUID -> list of entries of this stream.
+     */
+    public final void onNext(CorfuStreamEntries results) {
+
+        // If this update came before the client's full sync timestamp, ignore it.
+        if (results.getTimestamp().getSequence() <= clientFullSyncTimestamp.get()) {
+            return;
+        }
+
+        Set<String> tableNames =
+                results.getEntries().keySet().stream().map(schema -> schema.getTableName()).collect(Collectors.toSet());
+
+        if (tableNames.contains(REPLICATION_STATUS_TABLE)) {
+            Preconditions.checkState(results.getEntries().keySet().size() == 1,
+                "Replication Status Table Update received with other tables");
+            processReplicationStatusUpdate(results);
```

Comment:
QQ: Don't we need to check clientFullSyncPending and snapshotSyncInProgress before processReplicationStatusUpdate? Assuming that SyncPending and SyncinProgress are base checks before the processReplicationStatusUpdate.